File: Day-01/go/base2code/day01.go
package main

import (
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	content, _ := os.ReadFile("input.txt")
	lines := strings.Split(string(content), "\n")
	var left = make([]int, len(lines))
	var right = make([]int, len(lines))
	for i, line := range lines {
		split := strings.Split(line, "   ")
		left[i], _ = strconv.Atoi(split[0])
		right[i], _ = strconv.Atoi(split[1])
	}
	sort.Ints(left)
	sort.Ints(right)
	fmt.Println("Part 1:", part1(left, right))
	fmt.Println("Part 2:", part2(left, right))
}

func part1(left, right []int) int {
	distance := 0
	for i := 0; i < len(left); i++ {
		dis := math.Abs(float64(right[i] - left[i]))
		distance += int(dis)
	}
	return distance
}

func part2(left, right []int) int {
	cache := make(map[int]int)
	similarity := 0
	for _, leftnum := range left {
		if sim, ok := cache[leftnum]; ok {
			similarity += sim
		} else {
			count := 0
			for _, num := range right {
				if num == leftnum {
					count++
				}
			}
			cache[leftnum] = count * leftnum
			similarity += cache[leftnum]
		}
	}
	return similarity
}

File: Day-01/r/rainbowdashlabs/day01.R
lines <- readLines("input.txt")

left <- NULL
right <- NULL

for (line in lines) {
  parts <- strsplit(line, "[[:space:]]+")
  nums <- unlist(as.numeric(unlist(parts)))
  left <- c(left, nums[1])
  right <- c(right, nums[2])
}

left <- sort(left)
right <- sort(right)

distances <- NULL

for (i in seq_along(left)) {
  distances <- append(distances, abs(left[i] - right[i]))
}

sprintf("Part 1: %d", sum(distances))

right <- unlist(right)

res <- 0

for (num in left) {
  res <- res + num * sum(right == num)
}

sprintf("Part 2: %d", res)
